p2p/mainrpcserver.go

```diff
@@ -2,20 +2,118 @@ package p2p
 
 import (
 	"context"
+	"fmt"
+	"net"
 
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/phoreproject/synapse/beacon"
 	pb "github.com/phoreproject/synapse/pb"
+	"github.com/phoreproject/synapse/primitives"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/reflection"
 )
 
 type mainRPCServer struct {
+	chain *beacon.Blockchain
 }
 
 // NewMainRPCServer creates a new pb.MainRPCServer
-func NewMainRPCServer() pb.MainRPCServer {
-	return mainRPCServer{}
+func NewMainRPCServer(chain *beacon.Blockchain) pb.MainRPCServer {
+	return mainRPCServer{chain: chain}
 }
 
-func (server mainRPCServer) Test(ctx context.Context, message *pb.TestMessage) (*pb.TestMessage, error) {
+func (s mainRPCServer) Test(ctx context.Context, message *pb.TestMessage) (*pb.TestMessage, error) {
+	fmt.Printf("Received test message: %s\n", message.Message)
 	return &pb.TestMessage{
 		Message: message.Message + " === this is response",
 	}, nil
 }
+
+// SubmitBlock submits a block to the network after verifying it
+func (s mainRPCServer) SubmitBlock(ctx context.Context, in *pb.SubmitBlockRequest) (*pb.SubmitBlockResponse, error) {
```

Comment:
This RPC really only needs sync info (requesting/sending blocks [in bulk]), blockchain info (get current slot number, get current block number), and version info (minimum version to connect to, etc). Also, note that version info cannot be through the protocol ID because then newer clients would never connect to older clients.

Also, maybe state info would be a good idea. We could even implement fast syncing where we checkpoint certain slots, ask someone to send the state for that block check that against the block hash.